src/librustc_typeck/check/demand.rs

```diff
@@ -58,7 +60,34 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         if let Err(e) = self.try_coerce(expr, checked_ty, expected) {
             let origin = TypeOrigin::Misc(expr.span);
             let expr_ty = self.resolve_type_vars_with_obligations(checked_ty);
-            self.report_mismatched_types(origin, expected, expr_ty, e);
+            let mut diag = self.mismatched_types_diag(origin, expected, expr_ty, e.clone());
+
+            if let Node::NodeBlock(block) = self.tcx.map.get(self.tcx.map.get_parent_node(expr.id)) {
+                if let TypeError::Sorts(ref values) = e {
+                    if values.expected.is_nil() {
+                        // an implicit return to a method with return type `()`
+                        diag.span_label(expr.span,
+                                        &"possibly missing `;` here?");
+                        // get the current node's method definition
+                        if let Node::NodeExpr(item) = self.tcx.map
+                            .get(self.tcx.map.get_parent_node(block.id))
+                        {
+                            if let Node::NodeItem(func) = self.tcx.map
+                                .get(self.tcx.map.get_parent_node(item.id))
+                            {
```

Comment:
Take a look at the [latest code](https://github.com/rust-lang/rust/pull/36409/files#diff-1d1b0d29a2e8da97c6bfb6e364d920c7R4101) , where I am still against the block's parent to see wether it is a function or a trait's method for the recommendation. Let me know if it still looks too out of place.

I also changed the label to be consistent with the existing phrasing "consider removing this semicolon"/"consider adding a semicolon here".